### PR TITLE
fix: Fix <Link> typing

### PR DESCRIPTION
**Pull request by itajaja**
We could actually be more precise but it might still be too restrictive. this is the approach taken by [semantic-ui](https://github.com/Semantic-Org/Semantic-UI-React/blob/master/src/elements/Button/Button.d.ts#L16) and others

### CODE REVIEW

**Comment by taion** [comment]
that would be best. see also https://github.com/4Catalyzer/found/pull/241

**Comment by itajaja** [comment]
yeah, I have my reservations about it because I think that it actually doesn't support necessarily all cases. my assumption might be wrong, and if you empirically found this to work OK on rb and astroturf, I am happy to implement. A link would help 🙏

**Comment by jquense** [comment]
https://github.com/react-bootstrap/react-bootstrap/blob/master/types/components/helpers.d.ts is a good start. I think works well, when it fails it fails to allow any props so not a ton concern about it over this

**Comment by itajaja** [comment]
thanks, I'll do that

**Comment by itajaja** [comment]
@taion should I do it on top of your refactoring?

**Comment by taion** [comment]
well, do we like that refactoring? it is a breaking change

**Comment by itajaja** [comment]
they seem sensical changes, it's up to you if you want to pull the trigger :O

**Comment by taion** [comment]
ok i'll merge that PR tomorrow then

**Comment by taion** [comment]
@itajaja Were you going to update this?

**Comment by itajaja** [comment]
eh forgot. let me do it now

**Comment by itajaja** [comment]
@jquense actually, I am not sure how do to it without transforming LinkProps in a type, which will be problematic because we override it in userland. specifically, none of the componet in the bs repo that use the `As` trick define props of their own, while Link does

**Comment by itajaja** [comment]
in other words, this is not possible because ts complains:

```
An interface can only extend an object type or intersection of object types with statically known members
```

**Comment by itajaja** [comment]
ping, what do you guys think?

**Comment by itajaja** [comment]
uh never mind, it actually just works, I though declaration merging worked differently...

**Comment by itajaja** [comment]
eh, this doesn't work because `children` gets overridden in Props.component

```js
readonly props: Readonly<P> & Readonly<{ children?: ReactNode }>;
```

File: types/index.d.ts [hunk 1]
```diff
@@ -352,8 +352,34 @@ declare module 'found' {
...
+    ? JSX.IntrinsicElements[Tag]
+    : Tag extends React.SFC<infer Props>
+    ? Props & React.Attributes
+    : Tag extends React.ComponentClass<infer Props2>
+    ? (Tag extends new (...args: any[]) => infer Instance
+        ? Props2 & React.ClassAttributes<Instance>
+        : never)
+    : never;
+
+  type ReplaceProps<Inner extends React.ReactType, P> = Omit<
```
Comment (taion) [inline types/index.d.ts]: `React.ReactType` is deprecated... see the equivalent PR on 4c/layout for non-deprecated usages

File: types/index.d.ts [hunk 2]
```diff
@@ -352,8 +352,34 @@ declare module 'found' {
...
+  type LinkProps<As extends React.ReactType> = ReplaceProps<
+    As,
+    BaseLinkProps
+  > & { as?: As };
+
+  class Link<As extends React.ReactType = 'a'> extends React.Component<
+    LinkProps<As>
+  > {
     onClick: (event: React.SyntheticEvent<any>) => void;
+    readonly props: LinkProps<As>;
```
Comment (taion) [inline types/index.d.ts]: is this some magical thing for functional children?

note that the signature of the props passed to children is not the same as what `<Link>` takes; it's actually only

```ts
href: string
active: boolean
onClick: () => ...
```

File: types/example-app/CustomLink.tsx [hunk 3]
```diff
@@ -2,7 +2,7 @@ import * as React from 'react';
 
 interface Props {
   href: string;
-  active: boolean;
+  active?: boolean;
```
Comment (taion) [inline types/example-app/CustomLink.tsx]: `active` should always be there

File: types/index.d.ts (hunk 1 above)
Comment (itajaja) [reply types/index.d.ts]: I couldn't find a deprecation doc anywhere

File: types/index.d.ts (hunk 2 above)
Comment (itajaja) [reply types/index.d.ts]: not, this overrides `readonly props: Readonly<P> & Readonly<{ children?: ReactNode }>;` which has the wrong `children` type

File: types/example-app/CustomLink.tsx (hunk 3 above)
Comment (itajaja) [reply types/example-app/CustomLink.tsx]: it's not in the types, and it can't. the reason being that `activeFieldName` is customizable

Comment (taion) [reply types/example-app/CustomLink.tsx]: erk, do we need two of these, then?

when rendered as a function child, do we get the correct type of `active: boolean`?

File: types/index.d.ts (hunk 1 above)
Comment (taion) [reply types/index.d.ts]: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/a9ba077376adae5812a96ccdc38a61cba3a36caa/types/react/index.d.ts#L70-L72

does your IDE not show this?

File: types/example-app/CustomLink.tsx (hunk 3 above)
Comment (itajaja) [reply types/example-app/CustomLink.tsx]: yes

File: types/index.d.ts (hunk 1 above)
Comment (itajaja) [reply types/index.d.ts]: uh, yeah, not off the bat (need to click)

Comment (taion) [reply types/index.d.ts]: ping

File: types/index.d.ts [hunk 4]
```diff
@@ -352,8 +352,22 @@ declare module 'found' {
         }) => React.ReactNode);
   }
 
-  class Link extends React.Component<LinkProps> {
+  type ReplaceProps<Inner extends React.ElementType, P> = Omit<
+    React.ComponentProps<Inner>,
```
Comment (itajaja) [inline types/index.d.ts]: ComponentProps is nice :)

**Comment by taion** [comment]
@itajaja I went through and fixed the types a bit. Can you take another look?

**Comment by taion** [comment]
@jquense 	Can we get an extra opinion here on whether these typings make sense?

File: types/index.d.ts [hunk 5]
```diff
@@ -4,7 +4,7 @@ declare module 'found' {
   import * as React from 'react';
   import { Reducer, Store, StoreEnhancer } from 'redux';
 
-  type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
+  type Omit<T, K extends keyof any> = Pick<T, Exclude<keyof T, K>>;
```
Comment (itajaja) [inline types/index.d.ts]: Omit is upstream in TS and it behaves like that

File: types/index.d.ts [hunk 6]
```diff
@@ -342,33 +342,48 @@ declare module 'found' {
     onClick?: (event: React.SyntheticEvent<any>) => void;
   }
 
+  interface LinkInjectedProps {
```
Comment (itajaja) [inline types/index.d.ts]: shouldn't this be `ComponentPropsWithRef<'a'>`?

File: types/index.d.ts (hunk 5 above)
Comment (taion) [reply types/index.d.ts]: it extends `keyof any` i think

File: types/index.d.ts (hunk 6 above)
Comment (taion) [reply types/index.d.ts]: eh, this seems to work?

Comment (itajaja) [reply types/index.d.ts]: but DRY!

File: types/index.d.ts (hunk 5 above)
Comment (itajaja) [reply types/index.d.ts]: right. I mean, use the upstream.

Comment (taion) [reply types/index.d.ts]: these ts defs are supposed to work with v3.0. so for 3.0 through 3.4, we can't

**Comment by taion** [comment]
@jquense any thoughts here?

**Comment by taion** [comment]
ping

**Comment by itajaja** [comment]
@taion let's move on with this 👍, I think it's fair

**Pull request merged**